Sprint 6 solution

DmitryGavrilovvv opened the pull request.

DmitryGavrilovvv committed:
Рефакторинг кода менеджера истории:
Добавить HashMap с двусвязным списком
Добавить методы удаления и вставки в конец списка

Добавить юнит тест для проверки сохранности id у удалённых задач

Добавить юнит тест для проверки актуальности списка подзадач после удаления подзадачи

DmitryGavrilovvv committed:
Добавить тест для проверки изменения полей задачи в менеджере

DmitryGavrilovvv committed:
Удалить все неиспользованные импорты

DmitryGavrilovvv committed:
Исправить стиль кода

DmitryGavrilovvv committed:
Исправить некорректную работу удаления задачи из истории

DmitryGavrilovvv committed:
Исправить code style

YazykovMikhail commented on src/ru/yandex/javacource/gavrilov/schedule/manager/Node.java:
```diff
@@ -0,0 +1,72 @@
+
+import java.util.Objects;
+
+public class Node {
```
```
	public static class Node {
		private Task task;
		private Node prev;
		private Node next;

		public Node(Task task, Node prev, Node next) {
			this.task = task;
			this.prev = prev;
			this.next = next;
		}
```

YazykovMikhail replied:
и все

YazykovMikhail commented on src/ru/yandex/javacource/gavrilov/schedule/manager/InMemoryHistoryManager.java:
```diff
@@ -1,29 +1,99 @@
-    public static final int MAX_SIZE = 10;
-    private final List<Task> history = new ArrayList<>();
+    private final Map<Integer, Node> history = new HashMap<>();
 
```
А вот как раз тут стоит добавить 
	private Node first;
	private Node last;

YazykovMikhail commented on src/ru/yandex/javacource/gavrilov/schedule/manager/InMemoryHistoryManager.java:
```diff
@@ -1,29 +1,99 @@
 
     @Override
-    public void addTaskInHistory(Task task){
+    public void addTaskInHistory(Task task) {
```
А как же добавление в историю?

	public void add(Task task) {
		if (task == null) {
			return;
		}
		final int id = task.getId();
		remove(id);
		linkLast(task);
		history.put(id, tail);
	}

YazykovMikhail commented on src/ru/yandex/javacource/gavrilov/schedule/manager/InMemoryHistoryManager.java:
```diff
@@ -1,29 +1,99 @@
+        return tasks;
+    }
+
+    private void removeNode(Node node, Integer id) {
```
соответсвенно будет так:
```
	private void removeNode(Node node) {

		if (node.prev != null) {
			node.prev.next = node.next;
			if (node.next == null) {
				tail = node.prev;
			} else {
				node.next.prev = node.prev;
			}
		} else {
			head = node.next;
			if (head == null) {
				tail = null;
			} else {
				head.prev = null;
			}
		}
	}
```

YazykovMikhail commented on src/ru/yandex/javacource/gavrilov/schedule/manager/InMemoryHistoryManager.java:
```diff
@@ -1,29 +1,99 @@
+        Node.setTail(node);
+    }
+
+    private List<Task> getTasks() {
```
Идея правильная, но можно обойтись одним циклом 

```
	private List<Task> getTasks() {
		List<Task> tasks = new ArrayList<>();

		Node node = first;
		while (node != null) {
			tasks.add(node.task);
			node = node.next;
		}
		return tasks;
	}
```

YazykovMikhail commented on src/ru/yandex/javacource/gavrilov/schedule/manager/InMemoryHistoryManager.java:
```diff
@@ -1,29 +1,99 @@
+    }
+
+    @Override
+    public void remove(int id) {
```
Сейчас у тебя в методе 3 раза ищеться в харнилище. Можно 1
```
	public void remove(int id) {
		final Node node = history.remove(id);
		if (node == null) {
			return;
		}
		removeNode(node);
	}
```

DmitryGavrilovvv committed:
Рефакторинг кода
Убрать head и tail из Node, и добавить в InMemoryHistoryManager
Исправить логику методов getTasks, removeNode, remove, add
Переименовать метод addInHistoryManager в add

DmitryGavrilovvv committed:
Исправить код стайл
